src/ring_theory/ideal/basic.lean

```diff
@@ -303,6 +303,69 @@ end pi
 
 end ideal
 
+/-- A right ideal in a semiring `R` is an additive submonoid `s` such that
+`b * a ∈ s` whenever `b ∈ s`. If `R` is a ring, then `s` is an additive subgroup.  -/
+@[reducible] def right_ideal (R : Type u) [semiring R] := submodule Rᵐᵒᵖ R
+
+namespace right_ideal
+
+variables [semiring α] (I : right_ideal α) {a b : α}
+
+protected lemma zero_mem : (0 : α) ∈ I := I.zero_mem
+
+protected lemma add_mem : a ∈ I → b ∈ I → a + b ∈ I := I.add_mem
+
+variables (a)
+lemma mul_mem_right : b ∈ I → b * a ∈ I := I.smul_mem (mul_opposite.op a)
+variables {a}
+
+@[ext] lemma ext {I J : right_ideal α} (h : ∀ x, x ∈ I ↔ x ∈ J) : I = J :=
+submodule.ext h
+
+lemma sum_mem {ι : Type*} {t : finset ι} {f : ι → α} :
+  (∀c∈t, f c ∈ I) → (∑ i in t, f i) ∈ I := submodule.sum_mem I
+
+theorem eq_top_of_unit_mem
+  (x y : α) (hx : x ∈ I) (h : x * y = 1) : I = ⊤ :=
+eq_top_iff.2 $ λ z _, calc
+    z = (x * y) * z : by simp [h]
+  ... = x * (y * z) : mul_assoc x y z
+  ... ∈ I : I.mul_mem_right _ hx
+
+theorem eq_top_of_is_unit_mem {x} (hx : x ∈ I) (h : is_unit x) : I = ⊤ :=
+let ⟨y, hy⟩ := h.exists_right_inv in eq_top_of_unit_mem I x y hx hy
+
+theorem eq_top_iff_one : I = ⊤ ↔ (1:α) ∈ I :=
+⟨by rintro rfl; trivial,
+ λ h, eq_top_of_unit_mem _ _ 1 h (by simp)⟩
+
+theorem ne_top_iff_one : I ≠ ⊤ ↔ (1:α) ∉ I :=
+not_congr I.eq_top_iff_one
+
+/-- A right ideal is a left ideal of the opposite (semi)ring. -/
+@[simps] def to_ideal : ideal αᵐᵒᵖ :=
+{ carrier := mul_opposite.op '' I,
```

Comment:
I think it's a better ideal to define this via the preimage, as then membership doesn't introduce an existential.

Does this generalize nicely to `submodule R A → submodule R Aᵐᵒᵖ` ?

Reply:
Right of course, what was I thinking not using `preimage`. Thanks!

Reply:
What about the submodule generalization?

Reply:
> Does this generalize nicely to submodule R A → submodule R Aᵐᵒᵖ ?

I presume the implied setup is that `A` carries a multiplication and is also `R`-module and it's a generalisation since I am turning a term of `submodule Rᵐᵒᵖ R` into a term of type `submodule Rᵐᵒᵖ Rᵐᵒᵖ`.

I guess it does generalise in this sense but I'm a bit dubious about doing it because the only situation where I can think of where this is useful is if the multiplication that `A` carries is compatible with the its `R`-module structure, in the sense that it is an `R`-algebra and then in this case we want `R` to be commutative, and in that case, who cares about `right_ideal`s?

I'll do it if you think I should but I'm a bit dubious.

Reply:
Sure, its dubious enough (and the implementation trivial enough) that it's fine not to add it until we need it.